tracer/src/Datadog.Trace/DataStreamsMonitoring/SchemaTrackingCommon.cs

```diff
@@ -0,0 +1,69 @@
+// <copyright file="SchemaTrackingCommon.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System.Globalization;
+
+namespace Datadog.Trace.DataStreamsMonitoring;
+
+internal class SchemaTrackingCommon
+{
+    internal interface IDescriptorWrapper
+    {
+        string? Name { get; }
+
+        Schema? GetSchema();
+    }
+
+    /// <summary>
+    /// Get the current span, add some tags about the schema,
+    /// and once in a while, tag it with the json of the whole schema used
+    /// </summary>
+    internal static void EnrichActiveSpanWith(IDescriptorWrapper descriptor, string operationName, string schematype)
+    {
+        var tracer = Tracer.Instance;
+
+        var settings = tracer.Settings;
+        if (!settings.IsDataStreamsMonitoringEnabled)
+        {
+            return;
+        }
+
+        var activeSpan = (tracer.ActiveScope as Scope)?.Span;
+        if (activeSpan == null)
+        {
+            return;
+        }
+
+        if (activeSpan.GetTag(Tags.SchemaType) == schematype)
+        {
+            // we already instrumented this, we are most likely in a recursive call due to nested schemas.
+            return;
+        }
```

Comment:
there is a lot of re-entering calls that are caught here, especially on the write path (for avro). In protobuf, this is called for nested schemas, for avro it's called for all attributes.
I wonder if that can be a perf issue, and if we have a more effective solution at preventing re-instrumentation ?

Reply:
Hmm I guess the tricky part of this is that there isn't a unique "Tags" class for these so we can't do the quick lookup.
As that would be the quickest IMO

Unsure if it is worthwhile though. Do you know if it is a performance issue? It may not be worthwhile to spend time on optimizing it at the moment.